File: 2. Structural Design Patterns/Decorator/C#/DecoratorTemplate.cs
// Decorator Pattern – C# Template

namespace DecoratorTemplate
{
    interface Component { void Operation(); }
    class ConcreteComponent : Component { public void Operation() { /* base work */ } }

    abstract class BaseDecorator : Component
    {
        protected readonly Component wrappee;
        protected BaseDecorator(Component wrappee) { this.wrappee = wrappee; }
        public virtual void Operation() { wrappee.Operation(); }
    }

    class ConcreteDecoratorA : BaseDecorator
    {
        public ConcreteDecoratorA(Component c) : base(c) {}
        public override void Operation() { /* extra A */ base.Operation(); }
    }

    class ConcreteDecoratorB : BaseDecorator
    {
        public ConcreteDecoratorB(Component c) : base(c) {}
        public override void Operation() { /* extra B */ base.Operation(); }
    }

    class Program { static void Main(){ Component c = new ConcreteDecoratorB(new ConcreteDecoratorA(new ConcreteComponent())); c.Operation(); } }
}

File: 2. Structural Design Patterns/Decorator/Python/DecoratorTemplate.py
# Decorator Pattern – Python Template

from abc import ABC, abstractmethod

class Component(ABC):
    @abstractmethod
    def operation(self):
        pass

class ConcreteComponent(Component):
    def operation(self):
        return "ConcreteComponent operation"

class BaseDecorator(Component):
    def __init__(self, component):
        self.component = component
    
    def operation(self):
        return self.component.operation()

class ConcreteDecoratorA(BaseDecorator):
    def operation(self):
        return f"ConcreteDecoratorA({super().operation()})"

class ConcreteDecoratorB(BaseDecorator):
    def operation(self):
        return f"ConcreteDecoratorB({super().operation()})"

def main():
    component = ConcreteComponent()
    decorated = ConcreteDecoratorB(ConcreteDecoratorA(component))
    print(decorated.operation())

if __name__ == "__main__":
    main()
